pkg/operation/botanist/dns.go

```diff
@@ -15,329 +15,194 @@
-// Terraform variables which are prefixed with TF_VAR_.
-func (b *Botanist) GenerateTerraformRoute53VariablesEnvironment(purposeInternalDomain bool) (map[string]string, error) {
-	secret, err := b.getDomainCredentials(purposeInternalDomain, awsbotanist.AccessKeyID, awsbotanist.SecretAccessKey)
-	if err != nil {
-		return nil, err
+		b.Logger.Infof("Waiting for %q DNS provider to be ready... (status=%s, message=%s)", name, status, message)
+		return false, nil
+	}); err != nil {
+		return gardencorev1alpha1helper.DetermineError(fmt.Sprintf("Failed to create DNS provider for %q DNS record: %q (status=%s, message=%s)", name, err.Error(), status, message))
 	}
-	keyValueMap := map[string]string{
-		"ACCESS_KEY_ID":     awsbotanist.AccessKeyID,
-		"SECRET_ACCESS_KEY": awsbotanist.SecretAccessKey,
-	}
-	return terraformer.GenerateVariablesEnvironment(secret, keyValueMap), nil
+
+	return nil
 }
 
-// GenerateTerraformAlicloudDNSVariablesEnvironment generates the environment containing the credentials which
-// are required to validate/apply/destroy the Terraform configuration. These environment must contain
-// Terraform variables which are prefixed with TF_VAR_.
-func (b *Botanist) GenerateTerraformAlicloudDNSVariablesEnvironment(purposeInternalDomain bool) (map[string]string, error) {
-	secret, err := b.getDomainCredentials(purposeInternalDomain, alicloudbotanist.AccessKeyID, alicloudbotanist.AccessKeySecret)
-	if err != nil {
-		return nil, err
+func (b *Botanist) deleteDNSProvider(ctx context.Context, name string) error {
+	if err := b.K8sSeedClient.Client().Delete(ctx, &dnsv1alpha1.DNSProvider{ObjectMeta: metav1.ObjectMeta{Namespace: b.Shoot.SeedNamespace, Name: name}}); err != nil && !apierrors.IsNotFound(err) {
+		return err
 	}
 
-	keyValueMap := map[string]string{
-		"ACCESS_KEY_ID":     alicloudbotanist.AccessKeyID,
-		"ACCESS_KEY_SECRET": alicloudbotanist.AccessKeySecret,
-	}
+	ctx, cancel := context.WithTimeout(ctx, 2*time.Minute)
+	defer cancel()
 
-	return terraformer.GenerateVariablesEnvironment(secret, keyValueMap), nil
+	return kutil.WaitUntilResourceDeleted(ctx, b.K8sSeedClient.Client(), &dnsv1alpha1.DNSProvider{ObjectMeta: metav1.ObjectMeta{Namespace: b.Shoot.SeedNamespace, Name: name}}, 5*time.Second)
 }
 
-// GenerateTerraformCloudDNSVariablesEnvironment generates the environment containing the credentials which
-// Terraform variables which are prefixed with TF_VAR_.
-func (b *Botanist) GenerateTerraformCloudDNSVariablesEnvironment(purposeInternalDomain bool) (map[string]string, error) {
-	secret, err := b.getDomainCredentials(purposeInternalDomain, gcpbotanist.ServiceAccountJSON)
-	if err != nil {
-		return nil, err
-	}
-	project, err := gcpbotanist.ExtractProjectID(secret.Data[gcpbotanist.ServiceAccountJSON])
-	if err != nil {
-		return nil, err
+func (b *Botanist) deployDNSEntry(ctx context.Context, name, dnsName, target string) error {
```

Comment:
Why do we deploy `DNSEntry` directly? Isn't it better to use annotated services like in https://github.com/gardener/external-dns-management/blob/1135e4bbd3c5b1ac8282b8afee34e7a0be0b63d0/examples/service_with_dns.yaml#L5-L6

Reply:
Maybe in the future we can do that, yes, but for both the external and the ingress DNS record we anyway need to deploy the `DNSEntry` resource ourselves. The reason is that that there is no Ingress/Service in the seed that we could annotate. Consequently, to keep the code structure easy to understand I implement it this way for all three DNS entries.

Reply:
What about the kube-apiserver's LB?

```yaml
apiVersion: v1
kind: Service
metadata:
  annotations:
    service.beta.kubernetes.io/aws-load-balancer-backend-protocol: ssl
    service.beta.kubernetes.io/aws-load-balancer-connection-idle-timeout: "3600"
    service.beta.kubernetes.io/aws-load-balancer-healthcheck-healthy-threshold: "2"
    service.beta.kubernetes.io/aws-load-balancer-healthcheck-interval: "30"
    service.beta.kubernetes.io/aws-load-balancer-healthcheck-timeout: "5"
    service.beta.kubernetes.io/aws-load-balancer-healthcheck-unhealthy-threshold: "2"
    service.beta.kubernetes.io/aws-load-balancer-ssl-negotiation-policy: ELBSecurityPolicy-TLS-1-2-2017-01
    service.beta.kubernetes.io/aws-load-balancer-ssl-ports: "443"
    dns.gardener.cloud/dnsnames: "api.external.com,api.internal.some.corp"
    dns.gardener.cloud/ttl: "500"
  labels:
    app: kubernetes
    role: apiserver
  name: kube-apiserver
  namespace: foo
spec:
  ports:
  - name: kube-apiserver
    port: 443
    protocol: TCP
    targetPort: 443
  selector:
    app: kubernetes
    role: apiserver
  type: LoadBalancer
```

The `external-dns-management` already supports multiple hostnames per service:
https://github.com/gardener/external-dns-management/blob/63f97b9a09397defbe227081ae40c905184c0bdd/pkg/dns/source/dnsinfo.go#L52-L58

Reply:
Doesn't help, the external DNS record shall point to the internal one, and only the internal one shall point to the kube-apiserver LB.
Also, the ingress DNS record is deployed in the shoot (addons-nginx-ingress-controller service), here we don't have a DNS controller watching yet (will come later).

Reply:
So you would like to have 

`External DOMAIN` ---CNAME---> `Internal DOMAIN` ---A---> `LoadBalancer's IP`?

For this we must specify longer `CNAME` `TTL` as we don't want end-users to often do 2 DNS Queries instead of 1.

Reply:
Exactly, that's how it is today (even without external DNS but what Gardener creates via Terraform).